### ISSUE TITLE
`renv` should check its renv version in every command

### ISSUE THREAD

**Issue by vspinu** [rstudio/renv#285]
If on 2 git branches live different verisions of `renv` and the user moved from one branch to another without restarting the R session, the `renv` from the old git branch will be used throughout. This can lead to very confusing bugs and errors.

 I think it would be safe at least to throw a warning and recommend restarting the R session. But ideally renv should re-boot itself (if that's possible).

**Comment by vspinu** [comment]
In fact the old version of renv is loaded even after R session restart. I see the right version of renv in the `activate.R` script and renv.lock (-66) but the library loaded is with version -46. `renv::upgrade` helps.

**Comment by kevinushey** [comment]
The current behavior is that `renv` will load whatever version is currently available within the project library:

https://github.com/rstudio/renv/blob/4f5f7e83fab8947b0ea2139372ba1385e766ddb4/renv/activate.R#L36-L57

This was intentional, as users were confused that calling e.g. `install.packages("renv")` within a project did not automatically "update" the version of `renv` used to manage a project. However, the activate script should probably warn (or message) when the version recorded in the activate script does not match the one in the project library.

This also does imply that if you're trying to maintain different versions of `renv` + project libraries on different branches, you'll have to reinstall `renv` and restart to get the matching version of `renv`. But that seems like a relatively uncommon world to live in -- most users will likely just use the latest CRAN release of `renv`, and use the same version of `renv` across branches.

I think the simplest way to handle this right now would be to set a custom library path for your separate branches, e.g.

```
RENV_PATHS_LIBRARY=renv/library-branch
```

However, it'd probably cleaner to have an approach with e.g. renv profiles as you've brought up before.

**Comment by vspinu** [comment]
> you're trying to maintain different versions of renv + project libraries

Well, not intentionally. `renv` is under fast development and I am keeping up-todate daily, but the production lacks behind for obvious reasons. I also tend to move between branches a lot because we have conflicting behavior due to different versions o thef packages. Renv was exceptionally helpful to track down the issues. 

> However, the activate script should probably warn (or message) when the version recorded in the activate script does not match the one in the project library.

Would be helpful indeed. I would say startup should also warn the discrepancy between renv.lock and the library. At this moment the burden to track different moving parts is a bit too much on the user side I would say. It's too easy to overlook package version mismatch (not just renv) when moving between branches, for example.

**Comment by kevinushey** [comment]
> Well, not intentionally. renv is under fast development and I am keeping up-todate daily, but the production lacks behind for obvious reasons. I also tend to move between branches a lot because we have conflicting behavior due to different versions o thef packages. Renv was exceptionally helpful to track down the issues.

Right. And I should say, thank you for tracking the development versions so closely and providing so much feedback. :-)

> Would be helpful indeed. I would say startup should also warn the discrepancy between renv.lock and the library. At this moment the burden to track different moving parts is a bit too much on the user side I would say. It's too easy to overlook package version mismatch (not just renv) when moving between branches, for example.

This sounds reasonable to me as well, although I do want to be careful around how "noisy" `renv` is. At least with Packrat, I think we were a bit too overbearing (especially with RStudio integration) in how often we stole attention from the user for things that didn't necessarily need to be resolved immediately.

A one-line startup message is probably appropriate, though.

**Comment by kevinushey** [comment]
I think the one-line startup message here should be sufficient in at least alerting users that the `renv` version in the library is no longer in sync with the lockfile + activate script, and so should be updated.

Feel free to re-open if you think there's something else I should address here.

**Comment by vspinu** [comment]
I just got 

```
renv autoloader requested version 0.8.3-73 but 0.8.3-66 was loaded instead 
```
It's a bit confusing. First, "autoloader" means the "activate script", right? Second, it looks like "renv" itself requested v-73, but then loaded v-66 instead and then also complained about that. Third, there is no actionable information here. What the user should or could do in this case?

**Comment by kevinushey** [comment]
Right, this could be better worded. Maybe:

```
renv x.y.z was loaded from project library, but z.y.x is recorded in lockfile.
Use `renv::restore(packages = "renv")` to install renv z.y.x.
Use `renv::snapshot()` to update the lockfile with renv x.y.z.
```

**Comment by kevinushey** [comment]
This also provides a bit more motivation for having a means to snapshot a single package, as you've suggested before -- it'd be nice to offer a recommendation that allowed one to update only `renv` in the lockfile (+ activate script) in this case.

**Comment by vspinu** [comment]
Thanks! This is perfect. Fixed in 15d852bc4ac7515.
